### PR TITLE
feat: api events

### PR DESCRIPTION
**Pull request by Daanvdplas**
Added events to contract library and pallet, including tests for the pallet.

Note: the weight for increase allowance has changed due to an extra storage read.

The events in the contract library will have to be demonstrated in an example contract. This will have to be done in a different PR.

### CODE REVIEW

File: pallets/api/src/fungibles/mod.rs
Comment (chungquantin) [inline pallets/api/src/fungibles/mod.rs:328]:
```suggestion
				id.clone().into(),
```

File: pallets/api/src/fungibles/mod.rs
Comment (chungquantin) [inline pallets/api/src/fungibles/mod.rs:365]:
```suggestion
			AssetsOf::<T>::set_metadata(origin, id.clone().into(), name.clone(), symbol.clone(), decimals)?;
```

**Review by chungquantin — CHANGES_REQUESTED** [review]

File: pallets/api/src/fungibles/mod.rs
Comment (peterwht) [inline pallets/api/src/fungibles/mod.rs:113]: `Approval` and `Transfer` are the only events per the PSP22.

File: pallets/api/src/fungibles/mod.rs
Comment (peterwht) [inline pallets/api/src/fungibles/mod.rs:124]: This does not seem useful to me. It is simply emitting the same event that pallet-assets is already emitting

File: pallets/api/src/fungibles/mod.rs
Comment (peterwht) [inline pallets/api/src/fungibles/mod.rs:133]: likewise

File: pallets/api/src/fungibles/mod.rs
Comment (peterwht) [inline pallets/api/src/fungibles/mod.rs:135]: likewise

File: pallets/api/src/fungibles/mod.rs
Comment (peterwht) [inline pallets/api/src/fungibles/mod.rs:146]: likewise

**Review by peterwht — CHANGES_REQUESTED** [review]
Unfortunately, I think this requires some additional thought.

Currently, the events are duplicates to the ones emitted from `pallet-assets`. From a contract-developer's point of view, why is that useful to them? Either way, they will have to perform the same indexing of events, the only difference is if they choose an event from `pallet-assets` or from Pop's api pallet.

I think realistically, these events should abide by the PSP22 standard, and only those should be returned.

We should think on what is actually valuable to the contract developers, and to us.

One idea:
- A contract developer may want to keep track of how their contract is used. So, we probably want to know when an API endpoint is called, which contract called it, and which use-case. The particular parameters may not be relevant as the underlying pallet can be indexed as well.

Rather, simply emitting events like: "pallet-api.psp22.create {contract_addr}` gives us and the contract developer knowledge of often their contract calls into the runtime.

**Comment by Daanvdplas** [comment]
> One idea:
A contract developer may want to keep track of how their contract is used. So, we probably want to know when an API endpoint is called, which contract called it, and which use-case. The particular parameters may not be relevant as the underlying pallet can be indexed as well.

Please create an issue for it as I'm not completely following you.

**Comment by Daanvdplas** [comment]
I agree that the events in the api pallet are perhaps unnecessary. However, in the contract library it might still be useful. We provide the events that developers "can" use when they are calling the apis for asset management. This immediately also provides some form of standard.

**Comment by peterwht** [comment]
> I agree that the events in the api pallet are perhaps unnecessary. However, in the contract library it might still be useful. We provide the events that developers "can" use when they are calling the apis for asset management. This immediately also provides some form of standard.

Yes, I agree. The ones in the contract are useful and should be kept!

File: pallets/api/src/fungibles/mod.rs
Comment (Daanvdplas) [reply pallets/api/src/fungibles/mod.rs:124]: I agree with you. Removed those events from the pallet.

Comment (Daanvdplas) [reply pallets/api/src/fungibles/mod.rs:124]: see refactor: remove duplicated events

File: pallets/api/src/fungibles/mod.rs
Comment (evilrobot-01) [inline pallets/api/src/fungibles/mod.rs:100]: Missing doc comments on the the event type itself.

File: pallets/api/src/fungibles/mod.rs
Comment (evilrobot-01) [inline pallets/api/src/fungibles/mod.rs:198]: `unlookup` returns a `Source` -> `spender_source` would be a better name.

File: pallets/api/src/fungibles/mod.rs
Comment (evilrobot-01) [inline pallets/api/src/fungibles/mod.rs:199]: `id_param` is not used when `Equal`, better to move it into the `Greater`/`Less` match arms to avoid the unnecessary clone when equal. Very low likelihood of the code path being run but an easy optimisation.

In the Greater case, it can be inserted directly inline into line 209 without the need for the variable.

File: pallets/api/src/fungibles/mod.rs
Comment (evilrobot-01) [inline pallets/api/src/fungibles/mod.rs:244]: Missing proof size in this simple read addition? Sadly should have a benchmark.

File: pallets/api/src/fungibles/mod.rs
Comment (evilrobot-01) [inline pallets/api/src/fungibles/mod.rs:281]: Naming as above.

File: pallets/api/src/fungibles/tests.rs
Comment (evilrobot-01) [inline pallets/api/src/fungibles/tests.rs:27]: Minor, but could just import the `System::assert_last_event` function to avoid the `System::` prefix in all the tests.

File: pop-api/src/v0/assets/fungibles.rs
Comment (evilrobot-01) [inline pop-api/src/v0/assets/fungibles.rs:61]: Missing doc comment for the public module. What is the point of this module? Some documentation will help users know why these events exist.

File: pop-api/src/v0/assets/fungibles.rs
Comment (evilrobot-01) [inline pop-api/src/v0/assets/fungibles.rs:92]: This and remaining events are missing topics.

File: pop-api/src/v0/assets/fungibles.rs
Comment (evilrobot-01) [reply pop-api/src/v0/assets/fungibles.rs:61]: I also think that the usage of these events should be demonstrated somewhere in this PR - e.g. unit or integration tests.

**Review by evilrobot-01 — CHANGES_REQUESTED** [review]
Looking great. There are some missing doc comments and some minor optimisations which can be done easily.

Simple weight addition is questionable and proof that the contract events work by inclusion in some tests would be ideal.

File: pallets/api/src/fungibles/tests.rs
Comment (Daanvdplas) [reply pallets/api/src/fungibles/tests.rs:27]: Because it is an associated function of the Pallet struct, rather than a free-standing function at the crate level, you cannot directly import it using frame_system::assert_last_event. Instead, it must be accessed through the Pallet struct.

The mock::System::assert_last_event import path assumes that System (which is an alias for frame_system::Pallet) has been defined in a mock module and that the assert_last_event function is accessible directly from this alias. However, the assert_last_event function is tied to the Pallet implementation for the frame_system module and is not a standalone function that can be imported directly like this.

File: pallets/api/src/fungibles/mod.rs
Comment (Daanvdplas) [reply pallets/api/src/fungibles/mod.rs:101]: See commit `refactor: incompatible event based on standard`

File: pallets/api/src/fungibles/mod.rs
Comment (Daanvdplas) [reply pallets/api/src/fungibles/mod.rs]: see commit `refactor: optimization pallet fungibles`

File: pallets/api/src/fungibles/mod.rs
Comment (Daanvdplas) [reply pallets/api/src/fungibles/mod.rs]: see commit `refactor: optimization pallet fungibles`

File: pallets/api/src/fungibles/mod.rs
Comment (Daanvdplas) [reply pallets/api/src/fungibles/mod.rs]: see commit `refactor: optimization pallet fungibles`

File: pop-api/src/v0/assets/fungibles.rs
Comment (Daanvdplas) [reply pop-api/src/v0/assets/fungibles.rs:92]: See commit `refactor: ink topics + docs to contract events`

File: pallets/api/src/fungibles/mod.rs
Comment (Daanvdplas) [reply pallets/api/src/fungibles/mod.rs]: Please see commit `fix: weight increase allowance`. This is a simple solution and not 100% accurate however before I put more time into this I would like to get your view on the approach that has been taken for `decrease_allowance`. Basically, we are reusing the weight function of `approve` for `increase_` and `decrease_allowance`. This is for all three we have to read state (`allowance`) and either call `approve_transfer`, `cancel_approval`, or both. As I think this is fine for now since we never underestimate weight, this does deserve attention before going to mainnet.

File: pop-api/src/v0/assets/fungibles.rs
Comment (Daanvdplas) [reply pop-api/src/v0/assets/fungibles.rs:61]: This will take me 2 hours. Will push something later tonight

**Review by chungquantin — APPROVED** [review]

File: pop-api/src/v0/assets/fungibles.rs
Comment (chungquantin) [inline pop-api/src/v0/assets/fungibles.rs:61]: Reviewed! The PR looks good to me. But I think we should have a documentation for this somewhere in the code (or in a guidebook) to mention that using these events in the contract would make the contract size larger. And if not needed, the developers should make use of the runtime event that we already exposed by default.

**Review by chungquantin — APPROVED** [review]

File: pop-api/src/v0/assets/fungibles.rs
Comment (chungquantin) [reply pop-api/src/v0/assets/fungibles.rs:61]: Link for reference: https://use.ink/basics/events#cost-of-using-events

File: pop-api/src/v0/assets/fungibles.rs
Comment (Daanvdplas) [reply pop-api/src/v0/assets/fungibles.rs:61]: EDIT: will do this tomorrow, created a separate issue for it to not delay this getting merged #181

Comment (Daanvdplas) [reply pop-api/src/v0/assets/fungibles.rs:61]: As for docs, see commit `docs: events contract`

File: pop-api/src/v0/assets/fungibles.rs
Comment (Daanvdplas) [reply pop-api/src/v0/assets/fungibles.rs:61]: docs: events contract

File: pop-api/src/v0/assets/fungibles.rs
Comment (evilrobot-01) [inline pop-api/src/v0/assets/fungibles.rs:61]: Great docs!

File: pop-api/src/v0/assets/fungibles.rs
Comment (evilrobot-01) [inline pop-api/src/v0/assets/fungibles.rs:134]: I think this is overkill, why would someone want the event indexed by the number of decimals?

File: pop-api/src/v0/assets/fungibles.rs
Comment (evilrobot-01) [inline pop-api/src/v0/assets/fungibles.rs:127]: Debatable, but probably okay.

File: pallets/api/src/fungibles/mod.rs
Comment (evilrobot-01) [inline pallets/api/src/fungibles/mod.rs]: Sadly I think this is problematic. I understand it was probably removed to conform to the standard, but does ommitting it effectively make it useless? The events are emitted from a pallet managing multiple assets, so without knowing which asset id had an approval, is there any point? A contract call might be able to correlate the resulting events based on the extrinsic events, but if contract makes two calls to the fungibles pallet to create an allowance, for two different assets, how can we tell from the events which value matches which asset?

**Review by evilrobot-01 — COMMENTED** [review]
Hmmm, not sure about the id removal from the events. See inline comments for more context.

**Comment by evilrobot-01** [comment]
A general heads up that all these PRs to the `api` branch do not seem to include any test coverage stats, meaning that there might still be some work to do before the branch is merged to `main`. I will take a look at that branch first to see why.

**Comment by evilrobot-01** [comment]
> A general heads up that all these PRs to the `api` branch do not seem to include any test coverage stats, meaning that there might still be some work to do before the branch is merged to `main`. I will take a look at that branch first to see why.

FYI - the `daan/api` branch is missing a number of commits from main, with the ci one perhaps the most important. I suggest getting this addressed soon so that you can get a feel for how much code is untested before assuming that the initial api work is done.

image

File: pallets/api/src/fungibles/mod.rs
Comment (Daanvdplas) [reply pallets/api/src/fungibles/mod.rs]: But perhaps this brings us back to this [comment](https://github.com/r0gue-io/pop-node/pull/153#pullrequestreview-2229006536). 

In short, reason for adding events to the fungible pallet:
- standard compliance - the problem is that the events are not useful though (see frank's comment). In addition, when we do add the `asset_id` the events become the same as in pallet assets (e.g. [Transferred](https://github.com/paritytech/polkadot-sdk/blob/819a5818284a96a5a5bd65ce67e69bab860d4534/substrate/frame/assets/src/lib.rs#L563))
- pallet usage - we have an indicator how often the pallet is used. More interesting is how often the API is used

Would it be oke to start without events in the pallet?

**Comment by Daanvdplas** [comment]
> FYI - the daan/api branch is missing a number of commits from main, with the ci one perhaps the most important. I suggest getting this addressed soon so that you can get a feel for how much code is untested before assuming that the initial api work is done.

On it right away

File: pop-api/src/v0/assets/fungibles.rs
Comment (Daanvdplas) [reply pop-api/src/v0/assets/fungibles.rs:134]: Didn't think here, thanks for pointing it out

File: pop-api/src/v0/assets/fungibles.rs
Comment (Daanvdplas) [reply pop-api/src/v0/assets/fungibles.rs]: Left it in "A good rule of thumb is to ask yourself if somebody might want to search for this topic." 

Could be.

File: pallets/api/src/fungibles/mod.rs
Comment (peterwht) [reply pallets/api/src/fungibles/mod.rs]: > Would it be oke to start without events in the pallet?

I would rather we start with something, even super simple events so we can be aware of if the API is being used on testnet.

Comment (Daanvdplas) [reply pallets/api/src/fungibles/mod.rs]: Alright I will then add the `asset id` back in the `Transfer` and `Approval` event

File: pallets/api/src/fungibles/mod.rs
Comment (peterwht) [inline pallets/api/src/fungibles/mod.rs:321]: Do we want to know if somebody creates an asset through Pop API? If yes, can please add a simple event here. Otherwise I am fine if you leave it as is.

**Review by peterwht — APPROVED** [review]
LGTM! Left a small comment, good with whatever you decide.

File: pallets/api/src/fungibles/mod.rs
Comment (evilrobot-01) [inline pallets/api/src/fungibles/mod.rs:286]: Why not have this check directly after the origin check? There are unnecessary reads which occur which dont factor into this check, so more efficient to return as early as possible.

File: pallets/api/src/fungibles/mod.rs
Comment (evilrobot-01) [inline pallets/api/src/fungibles/mod.rs:200]: This variable is called `weight` in `decrease_allowance` - suggest using the same name for consistency.

File: pallets/api/src/fungibles/mod.rs
Comment (evilrobot-01) [reply pallets/api/src/fungibles/mod.rs:321]: Good point.

File: pallets/api/src/fungibles/tests.rs
Comment (evilrobot-01) [inline pallets/api/src/fungibles/tests.rs:181]: For future, could probably just check if it is empty - less verbose.

**Review by evilrobot-01 — CHANGES_REQUESTED** [review]
Just one more optimisation and a variable rename for consistency. Peter has a good point in terms of emitting events to determine usage, so hopefully a quick add.

**Comment by peterwht** [comment]
[1124]

File: pallets/api/src/fungibles/mod.rs
Comment (Daanvdplas) [reply pallets/api/src/fungibles/mod.rs]: see [refactor: fungible pallet](https://github.com/r0gue-io/pop-node/pull/153/commits/3b3ae9fcf4de6115bc5a2ecb9a92a2ca2e0d0068)

File: pallets/api/src/fungibles/tests.rs
Comment (Daanvdplas) [reply pallets/api/src/fungibles/tests.rs]: I like it, thanks for pointing it out!

see [refactor: fungible pallet](https://github.com/r0gue-io/pop-node/pull/153/commits/3b3ae9fcf4de6115bc5a2ecb9a92a2ca2e0d0068)

File: pallets/api/src/fungibles/mod.rs
Comment (Daanvdplas) [reply pallets/api/src/fungibles/mod.rs]: Yes good one

[refactor: fungible pallet](https://github.com/r0gue-io/pop-node/pull/153/commits/3b3ae9fcf4de6115bc5a2ecb9a92a2ca2e0d0068)

File: pallets/api/src/fungibles/mod.rs
Comment (evilrobot-01) [inline pallets/api/src/fungibles/mod.rs:336]: I'd stick this as the first statement for consistency.

**Review by evilrobot-01 — COMMENTED** [review]
One last minor refactor then good for me.

File: pallets/api/src/fungibles/mod.rs
Comment (Daanvdplas) [reply pallets/api/src/fungibles/mod.rs]: see last commit

File: pallets/api/src/fungibles/mod.rs
Comment (evilrobot-01) [reply pallets/api/src/fungibles/mod.rs]: Strictly speaking, these are just dispachables and therefore callable from anywhere. There is no guarantee that the call actually came from a contract, but may be a high probability.

From an indexing perspective, if we can correlate a contract Called event with Created event with same transaction then we're fine. One would assume so.

**Review by evilrobot-01 — APPROVED** [review]

**Pull request merged**
